ntp.go

```diff
@@ -159,6 +162,87 @@ type QueryOptions struct {
 	TTL     int           // IP TTL to use for outgoing UDP packets, defaults to system default
 }
 
+// Checks if SNTP response is usable for time synchronisation using rules from
+// RFC5905 and other sanity checks.
+func (r *Response) IsUsable() bool {
+	// Reference Timestamp: Time when the system clock was last set or
+	// corrected. Semantics of this value seems to vary across NTP server
+	// implementations: it may be both NTP-clock time and system wall-clock
+	// time of this event. :-( So (T3 - ReferenceTime) is not true
+	// "freshness" as it may be actually NEGATIVE sometimes.
+	freshness := r.Time.Sub(r.ReferenceTime)
+
+	// (Lambda := RootDelay/2 + RootDispersion) check against MAXDISP (16s)
+	// is required as ntp.org ntpd may report sane other fields while
+	// giving quite erratic clock. The check is declared in packet() at
+	// https://tools.ietf.org/html/rfc5905#appendix-A.5.1.1.
+	lambda := r.RootDelay/2 + r.RootDispersion
+
+	// `r.RTT > 0` check is not included as it does not depend on the
+	// packet itself, but also depends on clock _speed_. It's indicator
+	// that local clock run faster than remote one, so (T4-T1) < (T3-T2),
+	// but it may be local clock issue.
+	// E.g. T1/T2/T3/T4 = 0/10/20/1 leads to RTT = -9s.
+
+	return r.Leap != LeapNotInSync && // RFC5905, packet()
+		0 < r.Stratum && r.Stratum < MaxStratum && // RFC5905, packet()
+		lambda < maxDispersion*time.Second && // RFC5905, packet()
+		!r.Time.Before(r.ReferenceTime) && // RFC5905, packet(), reftime <= xmt ~~ !(xmt < reftime)
+		freshness <= (1<<maxPoll)*time.Second && // ntpdate uses 24h as a heuristics instead of ~36h derived from MAXPOLL
+		ntpEpoch.Before(r.Time) && // sanity
+		ntpEpoch.Before(r.ReferenceTime) // sanity
+}
+
+// Returns single-packet estimate of the root synchronization distance. Some
+// SNTP clients limit check this value before using the response, e.g.
+// systemd-timesyncd uses 5.0s as an upper bound. See also
+// https://tools.ietf.org/html/rfc5905#appendix-A.5.5.2
+func (r *Response) RootDistance() time.Duration {
+	// RFC5905 suggests more strict check against _peer_ in fit(), that
+	// root_dist should be less than MAXDIST + PHI * LOG2D(s.poll).
+	// MAXPOLL is 17, so it is approximately at most (1s + 15e-6 * 2**17) =
+	// 2.96608 s, but MAXDIST and MAXPOLL are confugurable values in the
+	// reference implementation, so only MAXDISP check has hardcoded value
+	// in IsUsable().
+	//
+	// root_dist should also have following summands
+	// + Dispersion towards the peer
+	// + jitter of the link to the peer
+	// + PHI * (current_uptime - peer->uptime_of_last_update)
+	// but all these values are 0 if only single NTP packet was sent.
+	rtt := r.RTT
+	if rtt < 0 {
+		rtt = 0
+	}
+	return (rtt+r.RootDelay)/2 + r.RootDispersion
+}
+
+// Returns amount of causality violation between SNTP timestamps, it may be
```

Comment:
> two SNTP timestamps

I'd remove `two` as it's actually two pairs of timestamps. I'd also keep the reminder about leap seconds as leap second event is a possibility for significant Offset and large Offset paired with small RTT is a causality violation.

Reply:
Ok, how about this:

```go
// CausalityViolation returns a time duration representing the amount of
// causality violation between two sets of timestamps. It may be used as a
// lower bound on the time synchronization error. A leap second may
// contribute as much as 1 second of causality violation.
```

Reply:
I'd say
```
// CausalityViolation returns a time duration representing the amount of
// causality violation between two sets of timestamps. It may be used as a
// lower bound on current time synchronization error betwen local and NTP
// clock. A leap second may contribute as much as 1 second of causality violation.
```

I want to make it clear that we're estimating synchronization error of local clock WHILE sending SNTP packet, and we're NOT estimating synchronization error that MAY happen if we apply "knowledge" of the SNTP packet to tune local clock.

Reply:
ok looks good.